Use explicit item type map to determine field types and encodings

piehld opened the pull request:
This PR is intended to address a current limitation in the conversion of mmCIF to BCIF for some data items. 

Specifically, there's a bug when handling some uncommon data types such as `point_group` and `float_range`.
- `point_group` fails because it has "**int**" in the name (i.e., "po**int**_group")
- `float_range` fails because it has "**float**" in the name, so the code tries to encode it as a float, but really it's a string since the value contains a hyphen

Rather than having the code try to figure out what the type is and how to encode it based on substring matching, I've added a [current] complete list of possible types and what they should be encoded as. This is similar to the method being used by our Java encoder (e.g., see [here](https://github.com/rcsb/ciftools-java/blob/06e095fc0a97a3dba7e90b1345e7c90743f3d45b/src/main/java/org/rcsb/cif/schema/generator/SchemaGenerator.java#L522)). 

Additionally, I made a minor adjustment to the `__molstar_forced_ints` object, in which I changed it to be a dictionary rather than a list to allow for O(1) efficiency.

epeisach commented:
Okay - your code rolled back som submodules -- hence tests failing.

Is there a reason you are doing this?

Otherwise - fundamentally looks good...

epeisach commented:
I notice that you are using the term "field" in inMostarTypeHints -- maybe use the term "name" -- this would be consistent with PdbxContainers.py CifName class.

epeisach commented:
I reverted back the modules/cpp-cif-file to latest.
(git submodule update after you pull in changes to your branch)

I leave it to you to decide "field" vs "name" discussion.

epeisach commented:
You mentioned that this was detected by trying to convert public PDB files.  Perhaps a test case to show success?  I am not sure what the failure mode was with existing code and if it was detectable.  Or create a simple test script that outputs a minimal file - with point_group in it - and show it works - which would have failed before?

epeisach commented:
I have added a test - that would ail with current code and is fixed with your changes.  I needed to capture the output of logging and look for the cast error - which is no longer present.

piehld commented:
@epeisach 

> You mentioned that this was detected by trying to convert public PDB files. Perhaps a test case to show success? I am not sure what the failure mode was with existing code and if it was detectable. Or create a simple test script that outputs a minimal file - with point_group in it - and show it works - which would have failed before?

Good point—I've added additional test cases.

> I reverted back the modules/cpp-cif-file to latest.

Yes sorry about that, I had some outdated submodules that I accidentally pushed.

> I notice that you are using the term "field" in inMostarTypeHints -- maybe use the term "name" -- this would be consistent with PdbxContainers.py CifName class.

Thanks for the suggestion, I just updated it to be `name`.
